### PR TITLE
Unhandled exception in the background thread terminates the whole process

### PR DESCRIPTION
**Pull request by mrdavidlaing**
It happens here:  
https://github.com/bitpusher/Salient.ReliableHttpClient/blob/master/src/Salient.ReliableHttpClient/RequestController.cs#L239

### CODE REVIEW

**Comment by mrdavidlaing** [comment]
@bitpusher @fandrei - I apologise for derailed this issue by standing on my "bug reports require test cases" soapbox.  I still maintain that this issue would have been resolved quicker if it had been submitted along with a description or test that enabled everyone to understand what the actual issue is.

On that note, I believe that the actual issue we need to resolve is what to do with this exception edge case.

There are circumstances where [`request.Request.Abort()`](https://github.com/bitpusher/Salient.ReliableHttpClient/blob/master/src/Salient.ReliableHttpClient/RequestController.cs#L235-L240) will get executed.

What does `Request.Abort()` actually do?
1. Does it send an "abort" message to the HTTP server?
2. Does it clean up resources?
3. Does it throw an exception?
4. Do we actually need to call it?

From the perspective of a CIAPI.CS user, what is the correct behaviour at this point?
1. Should we swallow the exception?
2. Should we wrap the exception and rethrow with more description?

**Comment by fandrei** [comment]
I'm trying to reproduce this bug more often than once in a blue moon, but this one is particularly tough. However I know for sure, that this method _can_ throw exceptions, and that uncaught exception in background thread terminates the process.

File: src/Salient.ReliableHttpClient.UnitTests/RequestAbortFixture.cs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                    {
+                        if (!gate.WaitOne(request.Timeout))
+                        {
+//                            // #TODO: disassociate the HttpWebRequest from the result, abort it, complete the result with a timeout exception
+//                            Log.Error(string.Format("Aborting #{0} : {1} because it has exceeded timeout {2}",
+//                                                    request.Index, request.Request.RequestUri, request.Timeout));
+//                            request.Request.Abort();
+
+                            Console.WriteLine("Aborting request after timeout");
+                            request.Abort();
```
Comment (mrdavidlaing) [inline src/Salient.ReliableHttpClient.UnitTests/RequestAbortFixture.cs:80]: I can't figure out how request.Abort() can trigger the stack trace @fandrei is encountering.

My assumption was that calling request.Abort() triggered an exception; but that doesn't seem to be what's going on.

@bitpusher, @fandrei  - can you write some code that causes an exception to occur?

Comment (fandrei) [inline src/Salient.ReliableHttpClient.UnitTests/RequestAbortFixture.cs:80]: This method certainly doesn't always throw exception. Only in some specific circumstances.

**Comment by mrdavidlaing** [comment]
@fandrei What does `and that uncaught exception in background thread terminates the process` mean?

Which process?  Could you put together an example of what you mean?

**Comment by fandrei** [comment]
The process executing that code. http://en.wikipedia.org/wiki/Process_(computing)
You can put "throw Exception()" instead of request.Abort() in your sample, for instance.

**Comment by fandrei** [comment]
https://gist.github.com/fandrei/a86c463625e3820f44d9

**Comment by mrdavidlaing** [comment]
@bitpusher Please could you give @fandrei push access to this repo, so he can contribute code directly

File: src/TestConsoleApplication/Program.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+81
+   at System.Threading.ThreadHelper.ThreadStart_Context(Object state)
+   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionCo
+ntext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
+   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, C
+ontextCallback callback, Object state, Boolean preserveSyncCtx)
+   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, C
+ontextCallback callback, Object state)
+   at System.Threading.ThreadHelper.ThreadStart()
+     *
```
Comment (mrdavidlaing) [inline src/TestConsoleApplication/Program.cs:53]: @fandrei - Does this TestConsoleApplication simulate the "exception on background thread" behaviour you are talking about?

Comment (fandrei) [inline src/TestConsoleApplication/Program.cs:53]: @mrdavidlaing, looks the same.

**Comment by mrdavidlaing** [comment]
@fandrei, @bitpusher.  Hurrah!  We've reached a point where have some kind of test case that allows us to know we are all talking about the same thing.

So, you're a CIAPI.CS user, and you are writing code similar to this:

```
1:   var client = new AbortingHttpClient();
2:   var priceList = client.GetPriceList(marketId);
```

We know that in exceptional circumstances, line 2 can throw a `System.Net.WebException`

From the perspective of a CIAPI.CS user, what is the correct behaviour at this point?
1.  Should CIAPI.CS trap the exception and just return an empty priceList?
2.  Should CIAPI.CS trap the exception, add some more details as to what went wrong, and rethrow?

@fandrei - in your specific use case, what would be the most convienient behaviour?

@bitpusher - is it even possible to trap this exception inside CIAPI.CS?  And where does CIAPI.CS's retry functionality fit into the picture?

**Comment by fandrei** [comment]
Well, this test is artificial and it doesn't tell us why this exception happens in the first place. I will try to create an isolated test to see why this happens.

However, possible resolutions for this are:
1. Retry Abort (if its call is _necessary_ for the library to work correctly)
2. Suppress this exception and trigger necessary events (which should have been called by Abort) manually

In any case, the timeout exception should be passed to the caller of this request.

**Comment by bitpusher** [comment]
Here are some thoughts, which will be followed shortly by a solution.

Request.Abort() closes any active streams (request/response) and fires the async callback with a webexception.

So the actual exception happens within the asyncresult.

the attempt to simulate the case with 
     throw new WebException("Simulated exception from request.Abort");
simply throws another exception where there was none before and does not address the issue at hand.

The act of closing an active response stream, depending on the size of the content, can indicate to a server that there is no longer a client listening and allow abort on the server side but this cannot be assured.

see below for the relevant clr source.

```

        /* 
            Abort - Attempts to abort pending request,

            This calls into the delegate, and then closes any pending streams. 

            Input: none 

            Returns: none

        */ 
        public override void Abort()
        { 
#if DEBUG 
            using (GlobalLog.SetThreadKind(ThreadKinds.User /* | ThreadKinds.Async would be nice */)) {
#endif 
            Abort(null, AbortState.Public);
#if DEBUG
            }
#endif 
        }

        private void Abort(Exception exception, int abortState) 
        {
            GlobalLog.ThreadContract(ThreadKinds.Unknown, "HttpWebRequest#" + ValidationHelper.HashString(this) + "::Abort()"); 
            if (Logging.On) Logging.Enter(Logging.Web, this, "Abort", (exception == null? "" :  exception.Message));

            if(Interlocked.CompareExchange(ref m_Aborted, abortState, 0) == 0) // public abort will never drain streams
            { 
                GlobalLog.Print("HttpWebRequest#" + ValidationHelper.HashString(this) + "::Abort() - " + exception);

                m_OnceFailed = true; 
                CancelTimer();

                WebException webException = exception as WebException;
                if (exception == null)
                {
                    webException = new WebException(NetRes.GetWebStatusString("net_requestaborted", WebExceptionStatus.RequestCanceled), WebExceptionStatus.RequestCanceled); 
                }
                else if (webException == null) 
                { 
                    webException = new WebException(NetRes.GetWebStatusString("net_requestaborted", WebExceptionStatus.RequestCanceled), exception, WebExceptionStatus.RequestCanceled, _HttpResponse);
                } 

                try
                {
#if DEBUG 
                bool setResponseCalled = false;
                try 
                { 
#endif
                    // Want to make sure that other threads see that we're aborted before they set an abort delegate, or that we see 
                    // the delegate if they might have missed that we're aborted.
                    Thread.MemoryBarrier();
                    HttpAbortDelegate abortDelegate = _AbortDelegate;
#if DEBUG 
                    m_AbortDelegateUsed = abortDelegate == null ? (object) DBNull.Value : abortDelegate;
#endif 
                    if (abortDelegate == null || abortDelegate(this, webException)) 
                    {
                        // We don't have a connection associated with this request 

                        LazyAsyncResult chkConnectionAsyncResult = Async? null: ConnectionAsyncResult;
                        LazyAsyncResult chkReaderAsyncResult = Async? null: ConnectionReaderAsyncResult;

#if DEBUG
                        setResponseCalled = true; 
#endif 
                        SetResponse(webException);

                        if (chkConnectionAsyncResult != null)
                            chkConnectionAsyncResult.InvokeCallback(webException);
                        if (chkReaderAsyncResult != null)
                            chkReaderAsyncResult.InvokeCallback(webException); 
                    }
#if DEBUG 
                } 
                catch (Exception stressException)
                { 
                    t_LastStressException = stressException;
                    if (!NclUtilities.IsFatal(stressException)){
                        GlobalLog.Assert(setResponseCalled, "HttpWebRequest#{0}::Abort|{1}", ValidationHelper.HashString(this), stressException.Message);
                    } 
                    throw;
                } 
#endif 
                }
                catch (Exception ex) 
                {
                    if (NclUtilities.IsFatal(ex)) throw;
                }
                catch { } 
            }

            if(Logging.On)Logging.Exit(Logging.Web, this, "Abort", ""); 
        }

```

**Comment by bitpusher** [comment]
conclusion:

client libraries are expected to throw exceptions when exceptional conditions arise to allow the consuming code to respond as is appropriate for the case.

what you are seeing is by design.  

any time you access network resources you should expect that an exception can occur and respond appropriately.

so, @mrdavidlaing , the answer to your question is that nothing more is required, neither swallowing and returning empty results or providing more information. A ReliableHttpException:TimeoutException is already being thrown.

@fandrei, your background process must expect exceptions and respond appropriately.

```

   [TestFixture]
    public class TimeoutFixture
    {
        [Test]
        public void SimulateTimeout()
        {
            var gate = new AutoResetEvent(false);
            var testWeb = "Http://localhost:21266";
            var client = new ClientBase(new Serializer());

            client.BeginRequest(RequestMethod.GET, testWeb, "/DelayHandler.ashx", null, null, ContentType.TEXT, ContentType.TEXT, 
                TimeSpan.FromSeconds(1), 3000, 0, ar =>
                    {
                        Console.WriteLine("we are in the callback about to end the request");

                        // any time your code utilizes network resources you need to be ready to handle exceptions.

                        try
                        {
                            var response = client.EndRequest(ar);

                            Console.WriteLine("you will not see this line output, exception had already been triggered by .EndRequest()");
                        }
                        catch (TimeoutException ex)
                        {

                            Console.Write("our code properly caught a timeout exception:\r\n" + ex.ToString());
                        }

                        gate.Set();

                    }, null);


            gate.WaitOne(20000);
            client.Dispose();
        }
    }
}
```

```
we are in the callback about to end the request
our code properly caught a timeout exception:
The request was aborted: The request was canceled. - failed 1 times
System.Net.WebException
   at Salient.ReliableHttpClient.ReliableAsyncResult.End() in C:\dev\Salient.ReliableHttpClient\src\Salient.ReliableHttpClient\ReliableAsyncResult.cs:line 129
   at Salient.ReliableHttpClient.RequestController.EndRequest(ReliableAsyncResult result) in C:\dev\Salient.ReliableHttpClient\src\Salient.ReliableHttpClient\RequestController.cs:line 449
   at Salient.ReliableHttpClient.ClientBase.EndRequest(ReliableAsyncResult result) in C:\dev\Salient.ReliableHttpClient\src\Salient.ReliableHttpClient\ClientBase.cs:line 354
   at Salient.ReliableHttpClient.UnitTests.TimeoutFixture.<>c__DisplayClass1.<SimulateTimeout>b__0(ReliableAsyncResult ar) in C:\dev\Salient.ReliableHttpClient\src\Salient.ReliableHttpClient.UnitTests\TimeoutFixture.cs:line 30
System.Net.WebException: The request was aborted: The request was canceled.
   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult)
   at Salient.ReliableHttpClient.RequestInfo.CompleteRequest(IAsyncResult result) in C:\dev\Salient.ReliableHttpClient\src\Salient.ReliableHttpClient\RequestInfo.cs:line 79


```

the spirit of the current timeout implementation is correct. that said, the implementation is a hack that got forgotten and will be cleaned up, but the library behavior will remain unchanged.

**Comment by fandrei** [comment]
@bitpusher, my process handles exceptions on all CIAPI.CS calls it makes. But the problem is in a different place.
This exception happens in a thread created by RelialeHttpClient and goes uncaught to the root of that thread. When this happens, CLR terminates the executing process and this is impossible to prevent.

**Comment by bitpusher** [comment]
@fandrei,

The WebException that is thrown by the underlying HttpWebRequest is caught and converted to a ReliableHttpException:TimeoutException and stored in the ReliableAsyncResult. 

This exception is not thrown until your code calls EndRequest().

If you can demonstrate how your code is different from the case shown in https://github.com/bitpusher/Salient.ReliableHttpClient/blob/master/src/ConsoleTestApp/Program.cs I will be happy to help.

there is no need for exotic tools to simulate timeouts in this case. See https://github.com/bitpusher/Salient.ReliableHttpClient/tree/master/src/TestWeb  which is used by the test case above.

Please open a new issue and provide a test case so that i may help improve this library.

**Comment by fandrei** [comment]
@bitpusher, I don't see any way how exception thrown by Abort() method here can be handled. https://github.com/bitpusher/Salient.ReliableHttpClient/blob/master/src/Salient.ReliableHttpClient/RequestController.cs#L239

**Comment by bitpusher** [comment]
please read my explanation of Request.Abort() and the accompanying CLR source code. 

Request.Abort() does not throw an exception.

**Comment by fandrei** [comment]
It does - sometimes.

**Comment by bitpusher** [comment]
the only time an exception will come out of .Abort is if there is an underlying OutOfMemoryException or a StackOverflow exception at the root of the request being aborted and the private overload .Abort is being called by CLR.

our code calling public .Abort will never throw an exception. The exception is registered via LazyAsyncResult with the callback.

provide a repeatable example of the problem so that we can solve it.

**Comment by bitpusher** [comment]
correction: add  ThreadAbortException to that list. 

http://typedescriptor.net/name/members/DF5C5CF5A8810AFFC8AB9A2BDDA68A5B-System.Net.NclUtilities.IsFatal(Exception)

**Comment by fandrei** [comment]
@bitpusher, it's nearly impossible to create a "repeatable example" for the mandelbug. However, the stack trace in the second message obviously shows this method _can_ throw exception.

**Comment by mrdavidlaing** [comment]
@fandrei, you say:

> my process handles exceptions on all CIAPI.CS calls it makes. But the problem is in a different place.
> This exception happens in a thread created by RelialeHttpClient and goes uncaught to the root of that thread.
> When this happens, CLR terminates the executing process and this is impossible to prevent.

Are you saying that it is impossible to catch this exception?

**Comment by fandrei** [comment]
@mrdavidlaing, it's possible to catch it - only in the thread, where it happens.

**Comment by mrdavidlaing** [comment]
As demonstrated in https://github.com/bitpusher/Salient.ReliableHttpClient/pull/4, it isn't possible to trap the (edge case) exception being thrown by Request.Abort in the code calling the library.

This means that it isn't possible for the calling application to recover; since the CLR just terminates it.  Clearly not an a useful "feature" of the ReliableHttpClient from the client app perspective!

Given:
1.  In most timeout cases we do want to call `Request.Abort()` to enable the CLR to clean up resources;
2.  However, `Request.Abort()` does actually trigger an exception there is nothing we can actually do

The proposal behaviour is that ReliableHttpClient should just trap, log (at WARN level) & discard any exceptions that Request.Abort() throws.

@fandrei, would this proposed behaviour solve the issue from your perspective?

**Comment by fandrei** [comment]
@mrdavidlaing, yes.

**Comment by mrdavidlaing** [comment]
@bitpusher, looks like the solution we discussed is the way forward.

Please could you make it so, and if possible push out an updated CIAPI.CS so the fix can make it into the AppMetrics Monitors.

@fandrei, @bitpusher  - this was a really tricky issue to get resolved.  Thanks for sticking with me through the discussion this long.
